tensorflow/lite/kernels/activations.cc

```diff
@@ -576,42 +577,72 @@ TfLiteStatus SoftmaxPrepare(TfLiteContext* context, TfLiteNode* node) {
   TF_LITE_ENSURE_OK(context, GetInputSafe(context, node, 0, &input));
   TfLiteTensor* output;
   TF_LITE_ENSURE_OK(context, GetOutputSafe(context, node, 0, &output));
-  if (output->type == kTfLiteInt16) {
-    TF_LITE_ENSURE(context, input->type == kTfLiteInt8 ||
-                                input->type == kTfLiteUInt8 ||
-                                input->type == kTfLiteInt16);
-  } else {
-    TF_LITE_ENSURE_TYPES_EQ(context, input->type, output->type);
-  }
 
   TF_LITE_ENSURE(context, NumDimensions(input) >= 1);
 
+  if (input->type == kTfLiteUInt8 || input->type == kTfLiteInt8 ||
```

Comment:
Note that the checks here are similar to the ones in [TFLite Micro](https://github.com/tensorflow/tensorflow/blob/f66782631d276c390732a513cdef6ad91eb4e779/tensorflow/lite/micro/kernels/softmax_common.cc#L35). For int8 -> int8 and int16 -> int16, the scale and zero-point are restricted properly in [operator_property.cc](https://github.com/tensorflow/tensorflow/blob/f66782631d276c390732a513cdef6ad91eb4e779/tensorflow/lite/tools/optimize/operator_property.cc#L939), I'm not sure for the (u)int8 -> int16 and uint8 -> uint8 cases. We could eventually only do these checks when the reference kernels are called to avoid any potential backward incompatibility.

Reply:
Thank you for the information regarding the (u)int8 -> int16 path. I removed the checks and now only enforce the int8 -> int8 and int16 -> int16 cases.